### PR TITLE
Convert dependency objects to records

### LINKED ISSUE #22076: Convert dependency objects to records

### PR DESCRIPTION
**Pull request by roji**
Closes #22076

### CODE REVIEW

File: test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs [hunk 1]
```diff
@@ -639,119 +650,6 @@ private string ValidateConventionMethod(MethodInfo conventionMethod)
             return null;
         }
 
-        [ConditionalFact]
-        public virtual void Ensure_dependency_objects_are_consistent()
```
Comment (roji) [inline test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: @AndriySvyryd assuming you're OK with this removal, we can also remove Fixture.TryGetProviderOptionsDelegate which isn't being used anywhere else.

Comment (AndriySvyryd) [reply test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: Yes, but add a condition in the next test to check that the dependency object is a record

**Review by AndriySvyryd — APPROVED** [review]

File: test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs (hunk 1 above)
Comment (roji) [reply test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: Oh sorry, I hit merge before noticing this. Will do a fixup PR.

Comment (AndriySvyryd) [reply test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: Still finding ways to boost your commit count even with squashed being enforced.

Comment (roji) [reply test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: One must always find a way... :rofl: 

It's actually not entirely clear how to detect records in reflection. Will look into this.

Comment (smitpatel) [reply test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: Long way to go on commit count @roji  😉

Comment (roji) [reply test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:643]: https://github.com/dotnet/efcore/pull/23423#discussion_r528174774

**Pull request merged**